File: archive/canvas-api/src/main/java/com/instructure/canvasapi/api/FileFolderAPI.java
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi.api;

import com.instructure.canvasapi.model.CanvasContext;
import com.instructure.canvasapi.model.FileFolder;
import com.instructure.canvasapi.utilities.APIHelpers;
import com.instructure.canvasapi.utilities.CanvasCallback;
import com.instructure.canvasapi.utilities.LinkHeaders;

import retrofit.Callback;
import retrofit.RetrofitError;
import retrofit.client.Response;
import retrofit.http.DELETE;
import retrofit.http.GET;
import retrofit.http.Path;

public class FileFolderAPI extends BuildInterfaceAPI {

    interface FilesFoldersInterface {
        @GET("/{context_id}/folders/root")
        void getRootFolderForContext(@Path("context_id") long context_id, Callback<FileFolder> callback);

        @GET("/self/folders/root")
        void getRootUserFolder(Callback<FileFolder> callback);

        @GET("/folders/{folderid}/folders")
        void getFirstPageFolders(@Path("folderid") long folder_id, Callback<FileFolder[]> callback);

        @GET("/folders/{folderid}/files")
        void getFirstPageFiles(@Path("folderid") long folder_id, Callback<FileFolder[]> callback);

        @GET("/{fileurl}")
        void getFileFolderFromURL(@Path(value = "fileurl", encode = false) String fileURL, Callback<FileFolder> callback);

        @GET("/{next}")
        void getNextPageFileFoldersList(@Path(value = "next", encode = false) String nextURL, Callback<FileFolder[]> callback);

        @DELETE("/files/{fileid}")
        void deleteFile(@Path("fileid")long fileId, Callback<Response> callback);
    }

    /////////////////////////////////////////////////////////////////////////
    // API Calls
    /////////////////////////////////////////////////////////////////////////

    public static void getFirstPageFoldersRoot(CanvasContext canvasContext, final CanvasCallback<FileFolder[]> callback) {
        if (APIHelpers.paramIsNull(callback, canvasContext)) {
            return;
        }

        // Build a callback bridge. Use a CanvasCallback instead of a regular callback so we can use caching appropriately and
        // not have to make more API calls than necessary
        Callback<FileFolder> bridgeCallback = new CanvasCallback<FileFolder>(callback.getStatusDelegate()) {

            @Override
            public void cache(FileFolder fileFolder, LinkHeaders linkHeaders, Response response) {
                buildCacheInterface(FilesFoldersInterface.class, callback, null).getFirstPageFolders(fileFolder.getId(), callback);
            }

            @Override
            public void firstPage(FileFolder fileFolder, LinkHeaders linkHeaders, Response response) {

                //Handle if the fragment becomes detached. This isn't a CanvasCallback, so it's not automatic.
                if(callback == null || callback.getContext() == null) {return;}

                buildInterface(FilesFoldersInterface.class, callback, null).getFirstPageFolders(fileFolder.getId(), callback);
            }

            @Override
            public void failure(RetrofitError retrofitError) {
                callback.failure(retrofitError);
            }
        };

        // get from cache
        getRootFolder(canvasContext, true, callback, bridgeCallback);

        // get from network
        getRootFolder(canvasContext, false, callback, bridgeCallback);

    }


    public static void getFirstPageFilesRoot(CanvasContext canvasContext, final CanvasCallback<FileFolder[]> callback) {
        getFirstPageFilesRootChained(canvasContext, false, callback);
    }

    public static void getFirstPageFilesRootChained(CanvasContext canvasContext, boolean isCached, final CanvasCallback<FileFolder[]> callback) {
        if (APIHelpers.paramIsNull(callback, canvasContext)) {
            return;
        }

        // Build a callback bridge. Use a CanvasCallback instead of a regular callback so we can use caching appropriately and
        // not have to make more API calls than necessary
        Callback<FileFolder> bridgeCallback = new CanvasCallback<FileFolder>(callback.getStatusDelegate()) {

            @Override
            public void cache(FileFolder fileFolder, LinkHeaders linkHeaders, Response response) {
                buildCacheInterface(FilesFoldersInterface.class, callback, null).getFirstPageFiles(fileFolder.getId(), callback);
            }

            @Override
            public void firstPage(FileFolder fileFolder, LinkHeaders linkHeaders, Response response) {

                //Handle if the fragment becomes detached. This isn't a CanvasCallback, so it's not automatic.
                if(callback == null || callback.getContext() == null) {return;}

                buildInterface(FilesFoldersInterface.class, callback, null).getFirstPageFiles(fileFolder.getId(), callback);
            }

            @Override
            public void failure(RetrofitError retrofitError) {
                callback.failure(retrofitError);
            }
        };

        getRootFolder(canvasContext, isCached, callback, bridgeCallback);
    }

    private static void getRootFolder(CanvasContext canvasContext, boolean isCached, CanvasCallback callback, Callback<FileFolder> bridgeCallback) {
        FilesFoldersInterface foldersInterface = buildInterface(FilesFoldersInterface.class, callback, canvasContext);

        if (canvasContext.getType() == CanvasContext.Type.USER) {
            if(isCached) {
                buildCacheInterface(FilesFoldersInterface.class, callback, canvasContext).getRootUserFolder(bridgeCallback);
            } else {
                foldersInterface.getRootUserFolder(bridgeCallback);
            }
        } else {
            if(isCached) {
                buildCacheInterface(FilesFoldersInterface.class, callback, canvasContext).getRootFolderForContext(canvasContext.getId(), bridgeCallback);
            } else {
                foldersInterface.getRootFolderForContext(canvasContext.getId(), bridgeCallback);
            }
        }
    }

    public static void getFirstPageFolders(long folderid, CanvasCallback<FileFolder[]> callback) {
        if (APIHelpers.paramIsNull(callback) || folderid <= 0) {
            return;
        }

        buildCacheInterface(FilesFoldersInterface.class, callback, null).getFirstPageFolders(folderid, callback);
        buildInterface(FilesFoldersInterface.class, callback, null).getFirstPageFolders(folderid, callback);
    }

    public static void getFirstPageFiles(long folderid, CanvasCallback<FileFolder[]> callback) {
        if (APIHelpers.paramIsNull(callback) || folderid <= 0) {
            return;
        }

        buildCacheInterface(FilesFoldersInterface.class, callback, null).getFirstPageFiles(folderid, callback);
        buildInterface(FilesFoldersInterface.class, callback, null).getFirstPageFiles(folderid, callback);
    }

    public static void getFirstPageFilesChained(long folderid, boolean isCached, CanvasCallback<FileFolder[]> callback) {
        if (APIHelpers.paramIsNull(callback) || folderid <= 0) {
            return;
        }

        if(isCached) {
            buildCacheInterface(FilesFoldersInterface.class, callback, null).getFirstPageFiles(folderid, callback);
        } else {
            buildInterface(FilesFoldersInterface.class, callback, null).getFirstPageFiles(folderid, callback);
        }
    }

    public static void getNextPageFileFolders(String nextURL, CanvasCallback<FileFolder[]> callback) {
        if (APIHelpers.paramIsNull(callback, nextURL)) {
            return;
        }

        callback.setIsNextPage(true);
        buildCacheInterface(FilesFoldersInterface.class, callback, false).getNextPageFileFoldersList(nextURL, callback);
        buildInterface(FilesFoldersInterface.class, callback, false).getNextPageFileFoldersList(nextURL, callback);
    }

    public static void getFileFolderFromURL(String url, CanvasCallback<FileFolder> callback) {
        if (APIHelpers.paramIsNull(callback, url)) {
            return;
        }

        buildCacheInterface(FilesFoldersInterface.class, callback, null).getFileFolderFromURL(url, callback);
        buildInterface(FilesFoldersInterface.class, callback, null).getFileFolderFromURL(url, callback);
    }

    public static void deleteFile(long fileId, CanvasCallback<Response> callback) {
        if (APIHelpers.paramIsNull(callback)) {
            return;
        }

        buildInterface(FilesFoldersInterface.class, callback, null).deleteFile(fileId,callback);
    }
}

File: libs/canvas-api-2/src/main/java/com/instructure/canvasapi2/apis/FileFolderAPI.kt
/*
 * Copyright (C) 2017 - present Instructure, Inc.
 *
 *     Licensed under the Apache License, Version 2.0 (the "License");
 *     you may not use this file except in compliance with the License.
 *     You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *     Unless required by applicable law or agreed to in writing, software
 *     distributed under the License is distributed on an "AS IS" BASIS,
 *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *     See the License for the specific language governing permissions and
 *     limitations under the License.
 *
 */

package com.instructure.canvasapi2.apis

import androidx.annotation.WorkerThread
import com.instructure.canvasapi2.StatusCallback
import com.instructure.canvasapi2.builders.RestBuilder
import com.instructure.canvasapi2.builders.RestParams
import com.instructure.canvasapi2.models.*
import com.instructure.canvasapi2.utils.APIHelper
import com.instructure.canvasapi2.utils.DataResult
import retrofit2.Call
import retrofit2.http.*
import java.io.IOException
import java.util.*


object FileFolderAPI {

    interface FilesFoldersInterface {

        @GET("self/folders/root")
        fun getRootUserFolder(): Call<FileFolder>

        @GET("{contextId}/folders/root")
        fun getRootFolderForContext(@Path("contextId") contextId: Long): Call<FileFolder>

        @GET("{contextType}/{contextId}/folders/root")
        suspend fun getRootFolderForContext(
            @Path("contextId") contextId: Long,
            @Path("contextType") contextType: String,
            @Tag params: RestParams
        ): DataResult<FileFolder>

        @GET("folders/{folderId}")
        fun getFolder(@Path("folderId") folderId: Long): Call<FileFolder>

        @GET("folders/{folderId}")
        suspend fun getFolder(@Path("folderId") folderId: Long, @Tag restParams: RestParams): DataResult<FileFolder>

        @GET("courses/{courseId}/files/{folderId}")
        fun getCourseFile(@Path("courseId") contextId: Long, @Path("folderId") folderId: Long): Call<FileFolder>

        @GET("courses/{courseId}/files/{fileId}")
        suspend fun getCourseFile(@Path("courseId") contextId: Long, @Path("fileId") folderId: Long, @Tag params: RestParams): DataResult<FileFolder>

        @GET("users/self/files/{folderId}")
        fun getUserFile(@Path("folderId") folderId: Long): Call<FileFolder>

        @GET("users/self/files/{folderId}")
        suspend fun getUserFile(@Path("folderId") folderId: Long, @Tag params: RestParams): DataResult<FileFolder>

        @GET("folders/{folderId}/folders")
        fun getFirstPageFolders(@Path("folderId") folderId: Long): Call<List<FileFolder>>

        @GET("folders/{folderId}/folders")
        suspend fun getFirstPageFolders(@Path("folderId") folderId: Long, @Tag params: RestParams): DataResult<List<FileFolder>>

        @GET("folders/{folderId}/files?include[]=usage_rights")
        fun getFirstPageFiles(@Path("folderId") folderId: Long): Call<List<FileFolder>>

        @GET("folders/{folderId}/files?include[]=usage_rights")
        suspend fun getFirstPageFiles(@Path("folderId") folderId: Long, @Tag params: RestParams): DataResult<List<FileFolder>>

        @GET("{fileUrl}")
        fun getFileFolderFromURL(@Path(value = "fileUrl", encoded = true) fileURL: String): Call<FileFolder>

        @GET("{fileUrl}")
        suspend fun getFileFolderFromURL(@Path(value = "fileUrl", encoded = true) fileURL: String, @Tag params: RestParams): DataResult<FileFolder>

        @GET
        fun getNextPageFileFoldersList(@Url nextURL: String): Call<List<FileFolder>>

        @GET
        suspend fun getNextPageFileFoldersList(@Url nextURL: String, @Tag params: RestParams): DataResult<List<FileFolder>>

        @GET("{canvasContext}/files")
        fun searchFiles(@Path(value = "canvasContext", encoded = true) contextPath: String, @Query("search_term") query: String): Call<List<FileFolder>>

        @GET("{canvasContext}/files")
        suspend fun searchFiles(@Path(value = "canvasContext", encoded = true) contextPath: String, @Query("search_term") query: String, @Tag params: RestParams): DataResult<List<FileFolder>>

        @DELETE("files/{fileId}")
        fun deleteFile(@Path("fileId") fileId: Long): Call<FileFolder>

        @PUT("files/{fileId}?include[]=usage_rights")
        fun updateFile(@Path("fileId") fileId: Long, @Body updateFileFolder: UpdateFileFolder): Call<FileFolder>

        @PUT("files/{fileId}?include[]=usage_rights")
        suspend fun updateFile(@Path("fileId") fileId: Long, @Body updateFileFolder: UpdateFileFolder, @Tag params: RestParams): DataResult<FileFolder>

        @POST("folders/{folderId}/folders")
        fun createFolder(@Path("folderId") folderId: Long, @Body newFolderName: CreateFolder): Call<FileFolder>

        @DELETE("folders/{folderId}?force=true")
        fun deleteFolder(@Path("folderId") folderId: Long): Call<FileFolder>

        @PUT("folders/{folderId}")
        fun updateFolder(@Path("folderId") folderId: Long, @Body updateFileFolder: UpdateFileFolder): Call<FileFolder>

        @FormUrlEncoded
        @PUT("courses/{courseId}/usage_rights")
        fun updateUsageRights(@Path("courseId") courseId: Long, @FieldMap params: Map<String, @JvmSuppressWildcards Any>): Call<UsageRights>

        @GET("courses/{courseId}/content_licenses")
        fun getCourseFileLicenses(@Path("courseId") courseId: Long): Call<ArrayList<License>>

        @GET("courses/{courseId}/content_licenses")
        suspend fun getCourseFileLicenses(@Path("courseId") courseId: Long, @Tag restParams: RestParams): DataResult<List<License>>

        @GET("files/{fileNumber}?include=avatar")
        fun getAvatarFileToken(@Path("fileNumber") fileNumber: String): Call<FileFolder>

        @GET("files/{fileId}")
        suspend fun getFile(@Path("fileId") fileId: Long, @Tag params: RestParams): DataResult<FileFolder>
    }

    fun getFileFolderFromURL(adapter: RestBuilder, url: String, callback: StatusCallback<FileFolder>, params: RestParams) {
        if (APIHelper.paramIsNull(callback, url)) return
        // TODO: add pagination
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getFileFolderFromURL(url)).enqueue(callback)
    }

    @WorkerThread
    fun getFileFolderFromURLSynchronous(adapter: RestBuilder, url: String, params: RestParams): FileFolder? {
        if (APIHelper.paramIsNull(url)) return null
        //TODO: add pagination
        try {
            return adapter.build(FilesFoldersInterface::class.java, params).getFileFolderFromURL(url).execute().body()
        } catch (e: IOException) {
            e.printStackTrace()
        }

        return null
    }

    fun getRootFolderForContext(adapter: RestBuilder, canvasContext: CanvasContext, callback: StatusCallback<FileFolder>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        if (canvasContext.type == CanvasContext.Type.USER) {
            callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getRootUserFolder()).enqueue(callback)
        } else {
            callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getRootFolderForContext(canvasContext.id)).enqueue(callback)
        }
    }

    fun getFolder(folderId: Long, adapter: RestBuilder, params: RestParams, callback: StatusCallback<FileFolder>) {
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getFolder(folderId)).enqueue(callback)
    }

    fun getCourseFile(courseId: Long, folderId: Long, adapter: RestBuilder, params: RestParams, callback: StatusCallback<FileFolder>) {
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getCourseFile(courseId, folderId)).enqueue(callback)
    }

    fun getUserFile(folderId: Long, adapter: RestBuilder, params: RestParams, callback: StatusCallback<FileFolder>) {
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getUserFile(folderId)).enqueue(callback)
    }

    fun getFirstPageFolders(adapter: RestBuilder, folderId: Long, callback: StatusCallback<List<FileFolder>>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getFirstPageFolders(folderId)).enqueue(callback)
    }

    fun getFirstPageFiles(adapter: RestBuilder, folderId: Long, callback: StatusCallback<List<FileFolder>>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getFirstPageFiles(folderId)).enqueue(callback)
    }

    fun getNextPageFilesFolder(adapter: RestBuilder, nextUrl: String, callback: StatusCallback<List<FileFolder>>, params: RestParams) {
        if (APIHelper.paramIsNull(callback, nextUrl)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getNextPageFileFoldersList(nextUrl)).enqueue(callback)
    }

    fun searchFiles(adapter: RestBuilder, query: String, canvasContext: CanvasContext, callback: StatusCallback<List<FileFolder>>, params: RestParams) {
        val contextPath = canvasContext.toAPIString().substring(1) // Drop leading slash
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).searchFiles(contextPath, query)).enqueue(callback)
    }

    fun deleteFile(adapter: RestBuilder, fileId: Long, callback: StatusCallback<FileFolder>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).deleteFile(fileId)).enqueue(callback)
    }

    fun updateFile(fileId: Long, updateFileFolder: UpdateFileFolder, adapter: RestBuilder, callback: StatusCallback<FileFolder>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).updateFile(fileId, updateFileFolder)).enqueue(callback)
    }

    fun createFolder(folderId: Long, folder: CreateFolder, adapter: RestBuilder, callback: StatusCallback<FileFolder>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).createFolder(folderId, folder)).enqueue(callback)
    }

    fun deleteFolder(folderId: Long, adapter: RestBuilder, callback: StatusCallback<FileFolder>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).deleteFolder(folderId)).enqueue(callback)
    }

    fun updateFolder(folderId: Long, updateFileFolder: UpdateFileFolder, adapter: RestBuilder, callback: StatusCallback<FileFolder>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).updateFolder(folderId, updateFileFolder)).enqueue(callback)
    }

    fun updateUsageRights(courseId: Long, formParams: Map<String, Any>, adapter: RestBuilder, callback: StatusCallback<UsageRights>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).updateUsageRights(courseId, formParams)).enqueue(callback)
    }

    fun getCourseFileLicenses(courseId: Long, adapter: RestBuilder, callback: StatusCallback<ArrayList<License>>, params: RestParams) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getCourseFileLicenses(courseId)).enqueue(callback)
    }

    fun getAvatarFileToken(fileNumber: String, adapter: RestBuilder, params: RestParams, callback: StatusCallback<FileFolder>) {
        if (APIHelper.paramIsNull(callback)) return
        callback.addCall(adapter.build(FilesFoldersInterface::class.java, params).getAvatarFileToken(fileNumber)).enqueue(callback)
    }
}
